File: gpt-4-turbo/stage2/parallel_construct_num_gangs_clause.f90
! This code example is designed to test the OpenACC 'parallel construct num_gangs clause'.
! It initializes an array and uses an OpenACC parallel region to modify the array
! elements. The success of various operations indicates proper handling of the num_gangs clause.

module testsuite_helpers
  use openacc
  implicit none
contains
  logical function test1()
    implicit none
    integer, parameter :: N = 1024
    integer :: i
    integer, dimension(N) :: a
    integer :: num_gangs_tested
    integer :: errors = 0
    
    ! Initialize array
    a = 0
    
    ! Choose an arbitrary number of gangs for test purposes
    num_gangs_tested = 128
    
    ! Use OpenACC parallel construct with num_gangs clause
    !$acc parallel loop num_gangs(num_gangs_tested)
    do i = 1, N
      a(i) = i
    end do
    
    ! Check results -- Simplistic check assuming uniform distribution of work
    do i = 1, N
      if (a(i) /= i) errors = errors + 1
    end do
    
    ! Determine the outcome based on errors found
    if (errors == 0) then
      test1 = .false.
    else
      test1 = .true.
    end if
  end function test1
  
end module testsuite_helpers

program main
  use testsuite_helpers
  implicit none
  logical :: failed
  integer :: failcode, testrun
  
  failed = .false.
  failcode = 0
  
  do testrun = 1, 1   ! Assuming a singular test run for this example
    failed = failed .or. test1()
  end do
  
  if (failed) then
    failcode = failcode + 2 ** 0
  end if
  
  call exit(failcode)
end program main

File: gpt-4-turbo/stage2/parallel_construct_num_gangs_clause.cpp
#include <iostream>
#include <stdlib.h>     // For rand(), srand()
#include <openacc.h>    // For OpenACC API

// Simple seed for reproducibility
#define SEED 42
#define NUM_TEST_CALLS 10

#ifndef T1
// T1:parallel construct num_gangs clause,V:2.7-3.3
int test1() {
    int err = 0;
    srand(SEED);
    const int size = 1000;
    const int numGangs = 10;
    int a[size] = {0};

    // Parallel region with a specified number of gangs
    #pragma acc parallel num_gangs(numGangs) copy(a)
    {
        #pragma acc loop
        for (int i = 0; i < size; ++i) {
            a[i]++;
        }
    }

    // Check for correctness
    for (int i = 0; i < size; ++i) {
        if (a[i] != 1) {
            err++;
            break;
        }
    }

    return err;
}
#endif

int main() {
    int failcode = 0;
    int failed;
#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x) {
        failed += test1();
    }
    if (failed != 0) {
        failcode += (1 << 0);
    }
#endif
    return failcode;
}